### ISSUE TITLE
skip_control may be confusing (it doesn't skip)

### ISSUE THREAD

**Issue by james-stocks** [inspec/inspec#3760]
### 🎛 Description

A customer wants to temporarily skip some controls. They use `skip_control` in a wrapper profile but find these controls have _no_ result rather than being skip results.

I believe this is as intended - `skip_control` and `skip` are different things. And `skip_control` would be more clearly named `exclude_control` ?

### Workaround

Use an `only_if`in the control with an appropriate condition.
Override the control in a wrapper profile and call `skip`.

**Comment by james-stocks** [comment]
I believe this has been explained to the customer in question and they won't need an update on this; but I'll leave the issue open with the Question label - it might come up again.

**Comment by jerryaldrichiii** [comment]
Yup, I think this should be `exclude_control` and then we can discuss if there should be a `skip_control`.

https://github.com/inspec/inspec/blob/a4362e7a0798af2c3eac3287fc083f6dc0529021/lib/inspec/control_eval_context.rb#L179

**Comment by kevin-j-smith** [comment]
Hi, I have been looking for a possible solution to skipping a control inside of a control.  See the only_if solution, which your documentation states:
```txt
This shows how to allow skipping certain tests if conditions are not met, by using only_if. In this example the test will not be performed if redis-cli command does not exist. A optional message can say why it was skipped.
```

Does not actually report the control or mark it skip.  It just excludes it.   (inspec version 3.7.1).  As to the skip command with in a describe block:
```ruby
describe 'skip the rest of the tests...this does not' do
    skip 'The control still keeps executing' if some_check?
end
```

The only way for this to work is to wrap everything in a control structure but the logic could be too complex (ie looping through an array) and makes everything way less elegant.   The information in this issue around 'skip_control' would be nice to have as a escape mechanism within the control.  There for I could:
```ruby
control "V-X" do
    skip_control "User Account Administration attribute does not have packages defined: #{uaa}" if uaa[:packages].nil?
    skip_control "User Account Administration attribute does not have any packages defined: #{uaa}" if uaa[:packages].length == 0
    skip_control "User Account Administration attribute does not have config_files defined: #{uaa}" if uaa[:config_files].nil?
    skip_control "User Account Administration attribute does not have any config_files defined: #{uaa}" if uaa[:config_files].length == 0

    uaa[:packages].each do |serviceName|
      s = service(serviceName)
       skip_control "Service #{serviceName} is not installed"  if !s.installed?
       skip_control "Service #{serviceName} is not running" if !s.running?
       skip_control "Service #{serviceName} is not enabled"  if !s.enabled?
    end

    uaa[:config_files].each do |config|
        describe file(config) do
          it { should exist }
          its('type') { should eq :file }
    end
```

In the end, I wish to bypass any tests if "skip_control" is ever set and report the control in the report with the "skip_control" message.   It would also be acceptable if "skip_control" accepted a block and anything within would be output into the message as well.

**Comment by kevin-j-smith** [comment]
How can I get the whole control to return the status as skipped.   I can do things in a ruby control structures to get all results to have status skipped.  That is without having a profile skip tests.  I am looking for a programmatic way to skip tests during run time.

**Comment by james-stocks** [comment]
Hi @kevin-j-smith - 

You said that for `only_if`...

> Does not actually report the control or mark it skip. It just excludes it. (inspec version 3.7.1).

This should cause a skipped result; and it's a bug if it doesn't.
I just tried in 3.7.1

```
control 'Important compliance check' do
  only_if { false }

  describe 'a' do
    it { should eq 'a' }
  end
end
```

```
➜  ~ inspec exec my_test.rb
WARN: Unresolved specs during Gem::Specification.reset:
      bundler (>= 1.10)
WARN: Clearing out unresolved specs.
Please report a bug if this causes problems.

Profile: tests from my_test.rb (tests from my_test.rb)
Version: (not specified)
Target:  local://

  ↺  Important compliance check: Operating System Detection
     ↺  Skipped control due to only_if condition.


Profile Summary: 0 successful controls, 0 control failures, 1 control skipped
Test Summary: 0 successful, 0 failures, 1 skipped
```

**Comment by kekaichinose** [comment]
@james-stocks not sure if it'll help, but with the upcoming waivers feature we're working on, waiving a control would by default still evaluate the control so that a user could track their actual result. Additionally though, we'd be looking to add something like a `skip_run` switch to indicate if the control should actually be skipped, meaning not evaluated.

Would that solve the use case here?

**Comment by james-stocks** [comment]
I think this is an issue as long as the docs and code for making a wrapper profile have the `skip_control` method.
Possibly a docs change would be sufficient to document that `skip_control` is a separate concept (control gathering/selection) to the skipping in control execution.

**Comment by zenspider** [comment]
This ties to #4408.

**Comment by zenspider** [comment]
OK. As it stands, I've talked about this at length with @kekaichinose and @miah... I think this is the result of a bad design. Controls need to be evaled because their metadata is provided by method calls inside the body of the control (as opposed to keyword args to the control that would cost _nothing_). As such, controls allow for heavy handed resources to be at the top level. We could probably put many more bandaids in place to try to have the resources detect that the control is being skipped, but that's NOT THEIR JOB. The design should be improved so that a skipped resource costs nothing AND has its metadata. 

I'm un-assigning this because it is gonna have to be InSpec v5 level epic planning and that's up for @kekaichinose to deal with.

**Comment by james-stocks** [comment]
There's 2 issues here:
1. The original issue description is only concerned with results - the customer expected to see the skipped control appear in their results list, with a skipped result.
2. Later comments in this issue discuss the bigger problem (that I wasn't aware of at the time of filing the issue) that skipped controls are still evaluated.

We should split this issue into 2, or make it an epic to fix up the whole area of control skipping.
